p2p/pex/pex_reactor.go

```diff
@@ -166,11 +168,16 @@ func (r *Reactor) OnStart() error {
 	return nil
 }
 
-// OnStop implements BaseService.
-func (r *Reactor) OnStop() {
+// Stop implements BaseService.
+func (r *Reactor) Stop() error {
+	if err := r.BaseReactor.Stop(); err != nil {
+		return err
+	}
 	if err := r.book.Stop(); err != nil {
```

Comment:
We should stop the address book first.

Reply:
This in particular will block all other calls to `r.book`, since for saving the address book it acquires the lock.

Reply:
Actually, for clarity, closing the address book should be implemented in the `OnStop` method.

Reply:
>  This in particular will block all other calls to r.book, since for saving the address book it acquires the lock.

Every call to the address bock acquires a lock. This lock is also acquired by `Save()`. So while saving, other calls to the address book will hang. From this point, assuming that the address book is stopped, if we change something on the address book, this changes won't be persisted.